### ISSUE TITLE
Train in 2 steps instead of 4?

### ISSUE THREAD

**Issue by ibarrien** [ivanpanshin/SupCon-Framework#26]
Hi thanks for your neat work here. It looks like there are currently 4 proposed steps to train:
for stage 1: train.py, swa.py and for stage 2: train.py, swa.py. 

Is there a way to pre-train using supcon loss once, and classification training once, for a total of two steps?

**Comment by ivanpanshin** [comment]
Hi! Glad you liked it. 

Well, one thing - SWA is actually optional. So SWA is about averaging weights of NN from several checkpoints. Should add some robustness and maybe even improvements on metrics. But it's absolutely okay to remove that part. 

Another thing - you might not need the second step (second stage) at all. So depends on how you want to use the trained net, because everything you might do with NN head (dense) - it's possible to do with embeddings from first stage.

Not sure whether it answers your question. Did you mean whether it's possible to remove SWA or whether it's possible to run both stages (NN body with SupCon + NN head with CE) in a single run?

**Comment by ibarrien** [comment]
Thank you, that's helpful -- good to know swa (weight averaging of checkpoints) can be skipped. Is it generally known i) how necessary checkpoint-weight-averaging is wrt validation loss and ii) whether weighted averaging can be incorporated more directly in training (rather than as a separate component)? Feel free to close this issue in case these questions may be out of scope :-)

**Comment by ivanpanshin** [comment]
1) Well, the ideal scenario to use SWA is when you have several local optima and you apply SWA on those checkpoints. For example, you might have CLR (cyclical learning rate) and you get local optima when your LR gets very low at each cycle. However, usually you need to somehow monitor local optima automatically (you can do it by hand, just easier to automate it). The simplest way - take the best checkpoints based either on your validation metrics (accuracy, f1_score, AP, etc), or validation loss.

2) Yeah, absolutely. Under the hood SWA is extremely simple. Since your NN is simply a dictionary of weights, SWA takes several dicts and applies something like np.mean() on the corresponding weights (there are some tricks with regards to batch norm, but you can skip this for simplicity). In other words, it's very easy to add this inside the training script (I just didn't do that due to legacy reasons, I guess. In my other project I need SWA as a separate script).
